### PR TITLE
feat: SectionMessage component

### PR DESCRIPTION
**Pull request by elliotlaws**
Ticket https://atomlearningltd.atlassian.net/browse/ACQ-651
Designs https://www.figma.com/file/KjgQh7eTwEfJmH6e56k9mX/DS-Documentation?node-id=9491-80193&t=gV2p1USSCRHfC9B9-0

```tsx
<SectionMessage>
  <SectionMessage.Icon />
  <SectionMessage.Content>
    <Box>
      <SectionMessage.Title>Title message</SectionMessage.Title>
      <SectionMessage.Description>
        This is the description
      </SectionMessage.Description>
    </Box>
    <SectionMessage.Actions>
      <Button size="sm">Button</Button>
      <Button size="sm" appearance="outline">
        Button
      </Button>
    </SectionMessage.Actions>
  </SectionMessage.Content>
</SectionMessage>
```

image

## Implementation

- Compound component for extra flexibility 
- Uses `Dismissible` component under the hood 
- Elements wrapped in `SectionMessage.Content` will either wrap/display inline depending on available width.

### CODE REVIEW

**Review by MiguelGDLR — COMMENTED** [review]
Great work Elliot! I have left a couple small review comments [here](https://www.figma.com/file/KjgQh7eTwEfJmH6e56k9mX/DS-Documentation?node-id=11180%3A99130&t=0tZXkzHLlDYKwYzn-1)

File: lib/src/components/section-message/SectionMessage.tsx [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    color: '$warningText',
+    icon: Danger
+  },
+  error: {
+    bg: '$dangerLight',
+    color: '$dangerDark',
+    icon: Error
+  },
+  neutral: {
+    bg: '$tonal100',
```
Comment (LimeWub) [inline lib/src/components/section-message/SectionMessage.tsx:38]: Could you use the `grey` scale (should be the same levels as the `blue` one)

`tonal` is getting phased out :)

Comment (elliotlaws) [reply lib/src/components/section-message/SectionMessage.tsx]: Updated 👍

**Comment by elliotlaws** [comment]
> Great work Elliot! I have left a couple small review comments [here](https://www.figma.com/file/KjgQh7eTwEfJmH6e56k9mX/DS-Documentation?node-id=11180%3A99130&t=0tZXkzHLlDYKwYzn-1)

Thanks Miguel; I've pushed some updates based on your feedback if you want to check

File: lib/src/components/section-message/SectionMessageLayout.tsx [hunk 2]
```diff
@@ -0,0 +1,29 @@
+import React from 'react'
+
+import { Flex } from '../flex'
+
+export const SectionMessageContent = ({
+  css,
+  ...rest
+}: React.ComponentProps<typeof Flex>): JSX.Element => {
+  return (
```
Comment (LimeWub) [inline lib/src/components/section-message/SectionMessageLayout.tsx:9]: I don't think we can use `columnGap` and `rowGap` due to support. But I also doubt we should be doing this as a Flex. We could instead have that as a container block/Box (see my comment about changing API slightly above) and just do the spacings in the child elements w margins.

Alternatively to do spacings of children via this have a look at using `Stack`

File: lib/src/components/section-message/SectionMessageText.tsx [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  css,
+  ...rest
+}: React.ComponentProps<typeof Text>): JSX.Element => {
+  const { theme } = useSectionMessageContext()
+
+  return (
+    <Text
+      css={{
+        fontWeight: 600,
+        fontSize: '$md',
```
Comment (LimeWub) [inline lib/src/components/section-message/SectionMessageText.tsx:17]: Please use size prop from Text

File: lib/src/components/section-message/SectionMessageText.tsx [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+export const SectionMessageDescription = ({
+  css,
+  ...rest
+}: React.ComponentProps<typeof Text>): JSX.Element => {
+  return (
+    <Text
+      css={{
+        fontWeight: 400,
+        fontSize: '$sm',
```
Comment (LimeWub) [inline lib/src/components/section-message/SectionMessageText.tsx:36]: size prop :)
Do you need to overwrite fontWeight?

File: documentation/content/components.feedback.section-message.md [hunk 5]
```diff
@@ -6,8 +6,151 @@ links:
   showReportAnIssue: true
 tabs:
   - title: Code
-    content: This component has not been built yet. You can check Visual and Usage
-      for now.
+    content: >-
+      A section message displays a contextual feedback message in a
+      particular section of the page. They’re persistent and nonmodal. It can
+      include a ‘close’ button to be dismissed, allowing the user to either
```
[the remark this answers is not in the document]
Comment (elliotlaws) [reply documentation/content/components.feedback.section-message.md:12]: From an API point of view, I prefer `Close` as I associate it more as being a close button. 

It also matches what we have for other components i.e. `Sidedrawer.Close` and `Dialog.Close`

File: documentation/content/components.feedback.section-message.md [hunk 6]
```diff
@@ -6,8 +6,151 @@ links:
...
+
+
+      `SectionMessage.Actions` will always align to the right when contained within `SectionMessage.Content` and shown inline.
+
+      <CodeBlock 
+        live={true} 
+        preview={true} 
+        language={"tsx"}
+        code={`
+          <SectionMessage css={{ width: 500 }}>
```
[the remark this answers is not in the document]
Comment (elliotlaws) [reply documentation/content/components.feedback.section-message.md:129]: The Title and Description weren't wrapped in a `Box`. How would we handle the wrapping behaviour shown in the designs?

File: documentation/content/components.feedback.section-message.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (LimeWub) [reply documentation/content/components.feedback.section-message.md:12]: These close popups. 
A similar component being finalised rn (Banner) is calling them `BannerRegular.Dismiss` which is what I was suggesting to match with.

File: documentation/content/components.feedback.section-message.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (LimeWub) [reply documentation/content/components.feedback.section-message.md:129]: You mean this:
Screenshot 2023-04-26 at 11 34 38
 vs
Screenshot 2023-04-26 at 11 34 51

?

**Comment by MiguelGDLR** [comment]
@elliotlaws All good from my side! great work 👊🏻

File: documentation/content/components.feedback.section-message.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (kerryambt) [reply documentation/content/components.feedback.section-message.md:12]: I think I agree with myrto here, there's a definite difference between closing a popup and dismissing something. we a) say the word a few words after close here, and b) use a dismissable component under the hood. I don't think close would be confusing but I think dismiss would be more accurate.

File: documentation/content/components.feedback.section-message.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (kerryambt) [reply documentation/content/components.feedback.section-message.md:129]: if you always needed title and description to be inside a box couldn't you split `children` into `title`, `description`, and `rest` by checking component types in SectionMessage.Content, and then output the required ones inside a box? or can we give that box a name in the API that makes it fit in more if that's not the case (SectionMessage.ContentBody or something)? I am not a fan of APIs that expect you to use unrelated components like this, just seems like it's opening up people forgetting it and getting unexpected results.

File: lib/src/components/section-message/SectionMessage.tsx [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+import {
+  SectionMessageActions,
+  SectionMessageContent
+} from './SectionMessageLayout'
+import {
+  SectionMessageDescription,
+  SectionMessageTitle
+} from './SectionMessageText'
+
+export const THEMES = {
```
Comment (kerryambt) [inline lib/src/components/section-message/SectionMessage.tsx:21]: if we're exporting a const can we give it a less generic name? `SECTION_MESSAGE_THEMES`?

File: documentation/content/components.feedback.section-message.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (LimeWub) [reply documentation/content/components.feedback.section-message.md:129]: What do you think about this solution? @elliotlaws 
https://github.com/Atom-Learning/components/pull/512

I'm assuming icon and x don't need moving so why not position absolute.

File: lib/src/components/section-message/SectionMessageClose.tsx [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+import { ActionIcon } from '../action-icon'
+import { Box } from '../box'
+import { Dismissible } from '../dismissible'
+import { Icon } from '../icon'
+
+export const SectionMessageClose = ({
+  label = 'Dismiss',
+  ...rest
+}: React.ComponentProps<typeof ActionIcon>): JSX.Element => {
+  return (
```
Comment (kerryambt) [inline lib/src/components/section-message/SectionMessageClose.tsx:13]: minor but if we don't need anything before the return can we just go `() => ( ...`?

File: documentation/content/components.feedback.section-message.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (elliotlaws) [reply documentation/content/components.feedback.section-message.md:12]: Happy to change to `Dismiss`

File: documentation/content/components.feedback.section-message.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (elliotlaws) [reply documentation/content/components.feedback.section-message.md:129]: Thanks for putting that together @LimeWub, but it doesn't seem to cater for it being inline when we have enough space

image

File: lib/src/components/section-message/SectionMessage.tsx [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+SectionMessage.Title = SectionMessageTitle
+SectionMessage.Description = SectionMessageDescription
+SectionMessage.Icon = SectionMessageIcon
+SectionMessage.Close = SectionMessageClose
+SectionMessage.Content = SectionMessageContent
+SectionMessage.Actions = SectionMessageActions
+
+export { SectionMessage }
```
[the remark this answers is not in the document]
Comment (elliotlaws) [reply lib/src/components/section-message/SectionMessage.tsx:99]: I just prefer stylistically how this looks, and I think it's easier to reason about seeing the export being the final thing that happens in this file

File: documentation/content/components.feedback.section-message.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (LimeWub) [reply documentation/content/components.feedback.section-message.md:129]: What JSX do you use for that example?

[the remark this answers is not in the document]
Comment (elliotlaws) [reply documentation/content/components.feedback.section-message.md:129]:
```tsx
      <SectionMessage theme="warning">
        <SectionMessage.Icon />
        <SectionMessage.Content>
          <SectionMessage.Title>Title message</SectionMessage.Title>
          <SectionMessage.Description>
            This is the description
          </SectionMessage.Description>
          <SectionMessage.Actions>
            <Button size="sm">Button</Button>
            <Button size="sm" appearance="outline">
              Button
            </Button>
          </SectionMessage.Actions>
        </SectionMessage.Content>
        <SectionMessage.Close />
      </SectionMessage>
```

[the remark this answers is not in the document]
Comment (LimeWub) [reply documentation/content/components.feedback.section-message.md:129]: my guess: I think you're putting actions in the content; but I've moved actions to be next to the content in my jsx

[the remark this answers is not in the document]
Comment (LimeWub) [reply documentation/content/components.feedback.section-message.md:129]: oh I didn't need to guess - yep.
Change to 
```
<SectionMessage theme="warning">
        <SectionMessage.Icon />
        <SectionMessage.Content>
          <SectionMessage.Title>Title message</SectionMessage.Title>
          <SectionMessage.Description>
            This is the description
          </SectionMessage.Description>
</SectionMessage.Content>
          <SectionMessage.Actions>
            <Button size="sm">Button</Button>
            <Button size="sm" appearance="outline">
              Button
            </Button>
          </SectionMessage.Actions>
        
        <SectionMessage.Close />
      </SectionMessage>
```

[the remark this answers is not in the document]
Comment (elliotlaws) [reply documentation/content/components.feedback.section-message.md:106]: @kerryambt the API has been updated as @LimeWub  example above

**Review by LimeWub — APPROVED** [review]

**Pull request merged**